repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -106,20 +100,34 @@ class Session(interpreter: Interpreter)
     _statements.clear()
   }
 
+  private def changeState(newState: SessionState): Unit = {
+    synchronized {
+      _state = newState
+    }
+    stateChangedCallback(newState)
+  }
+
   private def executeCode(executionCount: Int, code: String): String = synchronized {
-    _state = SessionState.Busy()
+    changeState(SessionState.Busy())
+
+    def transitToIdle() = {
+      val executingLastStatement = executionCount == newStatementId.intValue() - 1
+      if (_statements.isEmpty || executingLastStatement) {
```

Comment:
Frankly say, if I'm a user of Livy interpreter, I will not rely on "busy" or "idle" state, what I want to differentiate is "running" or not. I don't think Zeppelin really depends on that, not sure about Jupyter. Also as I mentioned before, such fine-grained status will increase the state management overhead for user (If user really want to differentiate busy state).